proposals/2024-11-25_changing_wal_format.md

```diff
@@ -0,0 +1,222 @@
+
+To achieve WAL versioning we also propose to start versioning the WAL, as [a whole format](https://github.com/prometheus/prometheus/blob/main/tsdb/docs/format/wal.md). While this is not strictly necessary for reading the data (at the moment), it is necessary to tell Prometheus what WAL format to write in.
+
+We propose the addition of a `meta.json` file in the wal directory, similar to [block meta.json](https://github.com/prometheus/prometheus/blob/5e124cf4f2b9467e4ae1c679840005e727efd599/tsdb/block.go#L171), with Version field set to `1` for the current format and `2` for new changes e.g. when we start to write [new records](https://github.com/prometheus/prometheus/pull/15467/files). No `meta.json` is equivalent to `{"version":1} `meta.json` file.
+
+We propose to also store the new WALs in separate directories e.g. `wal.v2`. Thanks to that the rewrite from one version to another is eventual and can be done segment by segment.
+The additional advantage of this way of versioning is that it's clear when your WAL fully migrated to a certain version.
+
+Finally, we propose that all Prometheus releases will contain the following table:
+
+| Data        | Supported | Writes |
+|-------------|-----------|--------|
+| WAL         | 1, 2      | 1      |
+| Block index | 2         | 2      |
+
+```
+The last revertable Prometheus version: v3.0 
+```
+
+See [alternatives](#alternatives) for other ideas.
+
+### Two-Fold Migration Strategy
+
+Given the following example:
+
+![twofold.png](../assets/2024-11-25_changing_wal_format/twofold.png)
+
+1. We release Prometheus X+1 version that supports both Y and Y+1 data but still writes Y.
+2. We release Prometheus X+2 version that supports both Y and Y+1 data, but now it writes new data as Y.
+
+Pros:
+* Users have a durable rollout path back and forth.
+* Dev has clarity on how to develop "breaking revert" changes.
+* In theory, it allows revert to X from X+2, by going through X+1 and ensuring all data was migrated (eventually) to Y version. In practice however that eventuality is long, or hard to discover.
+
+Cons:
+* It can catch users by surprise
+  * **Mitigation**: We also plan guide and documentation.
+* It takes time to rollout new changes if you want them fast.
+  * **Mitigation**: We plan to add a flag to opt-in sooner
+* It is a breaking change which shouldn't be made in a major release.
+* **Mitigation**: We accept that fact, given precedence and no other way to support both formats in the same time without major performance penalties.
+
+## Alternatives
+
+### Require LTS support
+
+We could add a variation to the [Two-Fold Migration Strategy](#two-fold-migration-strategy) where both X+1 and the last LTS (long time support version) before X+1, should understand Y+1 version -- before X+2 can write it by default.
+
+Pros:
+* This gives a bit more stability to users and less surprise.
+
+Cons:
+* We don't do LTS-es regularly
+* Is it safe or feasible to add Y+1 support to the old LTS code?
```

Comment:
I think it will take us some time to actually have a good version 2. There's a bunch of things to implement, so it's not far fetched to introduce it in the next LTS.

Reply:
We could introduce the flag right now as experimental. Use at own risk :) In fact maybe it has to be "experimental" all the time since you don't guarantee forward compatibility. Which would mean that:

There is no need to *set* a version through a flag. You need the `--version` flag to print the default version and the experimental version.

And to use the experimental version you'd set the `--enable-feature  wal-vX` . This would imply that we can break anything while the next version is experimental.

Reply:
> I think it will take us some time to actually have a good version 2. 

I like the idea of not forcing ourselves to only +1 version compatibility, but do the compatibility (and opt-in) as long as it's needed.

> There is no need to set a version through a flag. You need the --version flag to print the default version and the experimental version.

Prometheus --version? What do you mean? It prints only Prometheus versions. 

> And to use the experimental version you'd set the --enable-feature  wal-vX . This would imply that we can break anything while the next version is experimental.

Fair, we could do that instead of special flag. However it would be a very special feature flag which means you can't really go back to certain Prometheus version. How to not surprise users with that?

Reply:
I will rearrange proposal to switch into the ideas we discussed there, good points! I wonder if a new kind of feature flag would be useful here (e.g. `enable-storage-feature` or `enable-stateful-feature`).

Reply:
Ok, I thought through this. I don't think feature flag makes sense here -- added alternative option in the doc on why. Can you check? Let's open a new discussion. I added more details on the fact that we can keep "forward compatible, but not wriiting version" for more than one release in the proposal 🤗 

Thanks, good points!